File: dataflow.scala
// Lab in EDAN25 http://cs.lth.se/edan25/labs/

import scala.actors._
import java.util.BitSet;
import java.util.concurrent.Semaphore;
// LAB 2: some case classes but you need additional ones too.

case class Start();
case class Stop();
case class Ready();
case class Go();
case class Done();
case class Same();
case class Diff();
case class Change(in: BitSet);

class Random(seed: Int) {
        var w = seed + 1;
        var z = seed * seed + seed + 2;

        def nextInt() =
        {
                z = 36969 * (z & 65535) + (z >> 16);
                w = 18000 * (w & 65535) + (w >> 16);

                (z << 16) + w;
        }
}

class Controller(val cfg: Array[Vertex]) extends Actor {
  var started = 0;
  val begin   = System.currentTimeMillis();
  var sa           	= 0;
  var di           	= 0;

  def act() {
    react {
      case Ready() => {
        started += 1;
        // println("controller has seen " + started);
        if (started == cfg.length) {
        	// println("all started " + started);
        	sa = 0;
        	di = 0;
          for (u <- cfg){
          	// print("go ");
            u ! new Go;
            }
        }
        act();
      }
      case Same() => {
      	// println("Same " + (sa+di));
        sa = sa + 1;
        if(sa == cfg.length){
        	println("all done ");
        	for (u <- cfg){
            u ! new Stop;
            }
            this ! new Stop;
            Driver.mutex.release();

        }else if((sa+di) == cfg.length){
        	started = started -1;
        	this ! new Ready;

        	// for (u <- cfg){
         //    u ! new Start;
         //    }
        }
        act();
      }
      case Diff() => {
      	// println("Diff " + (sa+di));
        di = di + 1;
        if((sa+di) == cfg.length){
        	started = started -1;
        	this ! new Ready;
        	// for (u <- cfg){
         //    u ! new Start;
         //    }
        }
        act();
      }
      case Stop()  => { 

      }

    }
  }
}

class Vertex(val index: Int, s: Int, val controller: Controller) extends Actor {
  var pred: List[Vertex] = List();
  var succ: List[Vertex] = List();
  val uses               = new BitSet(s);
  val defs               = new BitSet(s);
  var in                 = new BitSet(s);
  var out                = new BitSet(s);
  var old               = new BitSet(s);
  var sn1            	= 0;
  var sn2            	= 0;
  val qwe 				= 12;
  def act() {
    react {
      case Start() => {
      	// if (index == qwe) println("start "+index+" sn "+sn1);
      	sn2 = sn1;
        controller ! new Ready;
        act(); 
      }
      case Change(in)=> {
      	// if (index == qwe) println("Change "+index+" sn "+sn2);
      	out.or(in);
        sn2=sn2-1;
        if(sn2==0){
          this ! new Done();
        }
      	act();
      }
      case Go() => {
      	// if (index == qwe) println("go "+index);
      	// sn2 = sn1;
      	for (v <- pred){
      		v ! new Change(in);
      	}
      	if(sn1==0){
          this ! new Done();
        }
        act();
      }
      case Done()  => {
      	// println("done "+index);
      	sn2 = sn1; 
        old = in;
    	in = new BitSet();
   	 	in.or(out);
    	in.andNot(defs);
   		in.or(uses);
   		if (in.equals(old)) {
      		controller ! new Same;
   		}else{
   			controller ! new Diff;
   		}
   		act();
      }
      case Stop()  => { 

      }
    }
  }

  def connect(that: Vertex)
  {
    //println(this.index + "->" + that.index);
    this.succ = that :: this.succ;
    that.pred = this :: that.pred;
    this.sn1 +=1;
  }

  def printSet(name: String, index: Int, set: BitSet) {
    System.out.print(name + "[" + index + "] = { ");
    for (i <- 0 until set.size)
      if (set.get(i))
        System.out.print("" + i + " ");
    println("}");
  }

  def print = {
    printSet("use", index, uses);
    printSet("def", index, defs);
    printSet("in", index, in);
    println("");
  }
}

object Driver {
  val rand    = new Random(1);
  var nactive = 0;
  var nsym    = 0;
  var nvertex = 0;
  var maxsucc = 0;
  var mutex = new Semaphore(0);

  def makeCFG(cfg: Array[Vertex]) {

    cfg(0).connect(cfg(1));
    cfg(0).connect(cfg(2));

    for (i <- 2 until cfg.length) {
      val p = cfg(i);
      val s = (rand.nextInt() % maxsucc) + 1;

      for (j <- 0 until s) {
        val k = cfg((rand.nextInt() % cfg.length).abs);
        p.connect(k);
      }
    }
  }

  def makeUseDef(cfg: Array[Vertex]) {
    for (i <- 0 until cfg.length) {
      for (j <- 0 until nactive) {
        val s = (rand.nextInt() % nsym).abs;
        if (j % 4 != 0) {
          if (!cfg(i).defs.get(s))
            cfg(i).uses.set(s);
        } else {
          if (!cfg(i).uses.get(s))
            cfg(i).defs.set(s);
        }
      }
    }
  }

  def main(args: Array[String]) {
    nsym           = args(0).toInt;
    nvertex        = args(1).toInt;
    maxsucc        = args(2).toInt;
    nactive        = args(3).toInt;
    val print      = args(4).toInt;
    val cfg        = new Array[Vertex](nvertex);
    val controller = new Controller(cfg);


    controller.start;

    println("generating CFG...");
    for (i <- 0 until nvertex)
      cfg(i) = new Vertex(i, nsym, controller);

    makeCFG(cfg);
    println("generating usedefs...");
    makeUseDef(cfg);

    println("starting " + nvertex + " actors...");

    for (i <- 0 until nvertex)
      cfg(i).start;

    for (i <- 0 until nvertex)
      cfg(i) ! new Start;

      mutex.acquire();

    if (print != 0)
      for (i <- 0 until nvertex)
        cfg(i).print;
  }
}

File: Dataflow.java
// Lab in EDAN25 http://cs.lth.se/edan25/labs/

import java.util.Iterator;
import java.util.ListIterator;
import java.util.LinkedList;
import java.util.BitSet;
import java.io.*;
import java.util.concurrent.Semaphore;


class Random {
	int	w;
	int	z;

	public Random(int seed) {
		w = seed + 1;
		z = seed * seed + seed + 2;
	}

	int nextInt() {
		z = 36969 * (z & 65535) + (z >> 16);
		w = 18000 * (w & 65535) + (w >> 16);

		return (z << 16) + w;
	}
}

class Vertex {
	int			index;
	boolean			listed;
	LinkedList<Vertex>	pred;
	LinkedList<Vertex>	succ;
	BitSet			in;
	BitSet			out;
	BitSet			use;
	BitSet			def;

	Vertex(int i) {
		index	= i;
		pred	= new LinkedList<Vertex>();
		succ	= new LinkedList<Vertex>();
		in	= new BitSet();
		out	= new BitSet();
		use	= new BitSet();
		def	= new BitSet();
	}

	void computeIn(LinkedList<Vertex> worklist) {
		int			i;
		BitSet			old;
		BitSet			ne;
		Vertex			v;
		ListIterator<Vertex>	iter;
		iter = succ.listIterator();
		while (iter.hasNext()) {
			v = iter.next();
			out.or(v.in);
		}
		old = in;
		in = new BitSet();
		in.or(out);
		in.andNot(def);
		in.or(use);
		if (!in.equals(old)) {
			iter = pred.listIterator();
			while (iter.hasNext()) {
				v = iter.next();
				if (!v.listed) {
					worklist.addLast(v);
					v.listed = true;
				}
			}

		}
	}

	public void print()	{
		try {
			PrintWriter writer = new PrintWriter(new FileOutputStream(new File("new.txt"), true));
			int	i;

			writer.print("use[" + index + "] = { ");
			for (i = 0; i < use.size(); ++i)
				if (use.get(i))
					writer.print("" + i + " ");
			writer.println("}");
			writer.print("def[" + index + "] = { ");
			for (i = 0; i < def.size(); ++i)
				if (def.get(i))
					writer.print("" + i + " ");
			writer.println("}\n");

			writer.print("in[" + index + "] = { ");
			for (i = 0; i < in.size(); ++i)
				if (in.get(i))
					writer.print("" + i + " ");
			writer.println("}");

			writer.print("out[" + index + "] = { ");
			for (i = 0; i < out.size(); ++i)
				if (out.get(i))
					writer.print("" + i + " ");
			writer.println("}\n");
			writer.close();
		} catch (Exception e) {
			e.printStackTrace();
		}
	}
}

class Dataflow {

	public static void connect(Vertex pred, Vertex succ) {
		pred.succ.addLast(succ);
		succ.pred.addLast(pred);
	}

	public static void generateCFG(Vertex vertex[], int maxsucc, Random r) {
		int	i;
		int	j;
		int	k;
		int	s;	// number of successors of a vertex.

		System.out.println("generating CFG...");

		connect(vertex[0], vertex[1]);
		connect(vertex[0], vertex[2]);

		for (i = 2; i < vertex.length; ++i) {
			s = (r.nextInt() % maxsucc) + 1;
			for (j = 0; j < s; ++j) {
				k = Math.abs(r.nextInt()) % vertex.length;
				connect(vertex[i], vertex[k]);
			}
		}
	}

	public static void generateUseDef(
	    Vertex	vertex[],
	    int	nsym,
	    int	nactive,
	    Random	r) {
		int	i;
		int	j;
		int	sym;

		System.out.println("generating usedefs...");

		for (i = 0; i < vertex.length; ++i) {
			for (j = 0; j < nactive; ++j) {
				sym = Math.abs(r.nextInt()) % nsym;

				if (j % 4 != 0) {
					if (!vertex[i].def.get(sym))
						vertex[i].use.set(sym);
				} else {
					if (!vertex[i].use.get(sym))
						vertex[i].def.set(sym);
				}
			}
		}
	}



	public static void main(String[] args) {
		int	i;
		int	nsym;
		int	nvertex;
		int	maxsucc;
		int	nactive;
		int	nthread;
		boolean	print;
		Vertex	vertex[];
		Random	r;

		r = new Random(1);

		nsym = Integer.parseInt(args[0]);
		nvertex = Integer.parseInt(args[1]);
		maxsucc = Integer.parseInt(args[2]);
		nactive = Integer.parseInt(args[3]);
		nthread = Integer.parseInt(args[4]);
		print = Integer.parseInt(args[5]) != 0;

		System.out.println("nsym = " + nsym);
		System.out.println("nvertex = " + nvertex);
		System.out.println("maxsucc = " + maxsucc);
		System.out.println("nactive = " + nactive);

		vertex = new Vertex[nvertex];

		for (i = 0; i < vertex.length; ++i)
			vertex[i] = new Vertex(i);

		generateCFG(vertex, maxsucc, r);
		generateUseDef(vertex, nsym, nactive, r);
		liveness(vertex, nthread);

		if (print)
			for (i = 0; i < vertex.length; ++i)
				vertex[i].print();
	}

	public static void liveness(Vertex vertex[], int nthread) {
		Vertex			u;
		LinkedList<Vertex>	worklist;
		long			begin;
		long			end;


		begin = System.nanoTime();
		worklist = new LinkedList<Vertex>();

		int partSize = vertex.length / nthread;
		LinkedList<livenessThread> t = new LinkedList<livenessThread>();

		for (int i = 0; i < nthread; i++) {
			worklist = new LinkedList<Vertex>();
			for (int j = i * partSize; j < (i + 1) * partSize; ++j) {
				worklist.addLast(vertex[j]);
				vertex[i].listed = true;
			}
			t.add(new livenessThread(worklist));
		}

		for (livenessThread livenessThread : t) {
			livenessThread.start();
		}
		try {
			for (livenessThread livenessThread : t) {
				livenessThread.join();
			}
		} catch (InterruptedException e) {
			e.printStackTrace();
		}

		end = System.nanoTime();

		System.out.println("T = " + (end - begin) / 1e9 + " s");
	}
}
class livenessThread extends Thread {

	Vertex			u;
	int			i;
	LinkedList<Vertex>	worklist;

	public livenessThread(LinkedList<Vertex> worklist) {
		this.worklist = worklist;
	}

	public void run() {

		System.out.println("computing liveness...");
		System.out.println("Worklist size: " + worklist.size());
		while (!worklist.isEmpty()) {
			u = worklist.remove();
			u.listed = false;
			u.computeIn(worklist);
		}

	}
}